### ISSUE TITLE
Rule proposal: `prefer-math-hypot`

### ISSUE THREAD

**Issue by yvele** [sindresorhus/eslint-plugin-unicorn#2079]
### Description

Enforces a convention of using [Math.hypot()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/hypot) instead of `Math.sqrt(v1*v1 + v2*v2 + vn*vn)` to make the calculation easier and faster. It also avoids overflow/underflow problems if the magnitude of your numbers is very large.

$\sqrt{a^2 + b^2 + … + n^2}$

### Fail

```js
Math.sqrt(a * a + b * b);
```

```js
Math.sqrt(a ** 2 + b ** 2);
```

```js
Math.sqrt(Math.pow(a, 2) + Math.pow(b, 2));
```

```js
Math.sqrt(a * a);
```

### Pass

```js
Math.hypot(a, b);
```

```js
Math.hypot(a, b, c);
```

```js
Math.hypot(a); // Equivalent to Math.abs(a)
```

### Additional Info

Not sure about auto fixing `Math.sqrt(a * a)` -> `Math.hypot(a)` as `Math.abs(a)` is better (but should be a dedicated rule?).

See also the [MDN documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/hypot#description):

> With one argument, `Math.hypot()` is equivalent to [Math.abs()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/abs). [Math.hypot.length](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/length) is 2, which weakly signals that it's designed to handle at least two parameters.

**Comment by fisker** [comment]
Can be part of  [prefer-modern-math-api](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/prefer-modern-math-apis.md)

**Comment by yvele** [comment]
Yeah. Although `Math.hypot` is not really "modern" https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/hypot#browser_compatibility

We could then also prefer `Math.abs(a)` over `Math.sqrt(a * a)` (and all other $\sqrt{a^2}$ variants)

**Comment by yvele** [comment]
Yeah.

~~Although `Math.hypot` is not really "modern" https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/hypot#browser_compatibility~~ (ok it's ES2015).

We could then also prefer `Math.abs(a)` over `Math.sqrt(a * a)` (with all $a^2$ JavaScript variants such as `a * a`, `Math.pow(a, 2)` and `a ** 2`)

But `Math.abs` is not "modern" maybe the rule should be renamed from `prefer-modern-math` to `prefer-specialized-math`? 🤔

**Comment by fisker** [comment]
I doubt if there are real-world cases using `Math.hypot(a)`, let's forget about it for now.

**Comment by yvele** [comment]
> I doubt if there are real-world cases using `Math.hypot(a)`, let's forget about it for now.

Ok.

But it should also be nice to enforce using `Math.abs(a)` (specialized math function) instead of `Math.sqrt(a * a)` (maybe I should open a dedicated issue?)

**Comment by yvele** [comment]
> I doubt if there are real-world cases using `Math.hypot(a)`, let's forget about it for now.

Ok.

But it should also be nice to enforce using `Math.abs(a)` (specialized math function) instead of `Math.sqrt(a * a)`.

Maybe I should open a dedicated issue for:

### Prefer `Math.abs(x)` over:

```js
Math.sqrt(x * x)
Math.sqrt(x ** 2)
Math.sqrt(Math.pow(x, 2))
```

```js
Math.hypot(x)
```

But `Math.abs` is not "modern" maybe the rule should be renamed from `prefer-modern-math-apis` to `prefer-specialized-math`? 🤔

**Comment by Primajin** [comment]
Hey friends, is it really faster to use `hypot` over `sqrt`? Do you have a benchmark at hand?

It looks like it's slower
<img width="612" alt="Screenshot 2023-07-26 at 11 25 33" src="https://github.com/sindresorhus/eslint-plugin-unicorn/assets/1742115/0af748d8-fe41-4a3d-a209-db7d093ed9c9">
 for me 🤔 (latest chrome arm64)

**Comment by fisker** [comment]
Did we mention about preference anywhere? I can't find.

**Comment by Primajin** [comment]
I just upgraded to v48 --> https://github.com/sindresorhus/eslint-plugin-unicorn/releases/tag/v48.0.0

And I think it's https://github.com/sindresorhus/eslint-plugin-unicorn/pull/2080 which changed that part of the code for me 😄
